src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/BuildUpToDateCheck.cs

```diff
@@ -0,0 +1,227 @@
+                    where TUnresolvedReference : IProjectItem, TResolvedReference
+                    where TResolvedReference : class, IReference
+                {
+                    Log($"Checking reference type '{name}'.");
+                    if (service != null)
+                    {
+                        foreach (var resolvedReference in await service.GetResolvedReferencesAsync())
+                        {
+                            var fullPath = await resolvedReference.GetFullPathAsync();
+                            Log($"Adding input reference ${fullPath}");
+                            inputs.Add(fullPath);
+                        }
+                    }
+                }
+
+                await CheckReferences(nameof(_configuredProject.Services.AssemblyReferences), _configuredProject.Services.AssemblyReferences);
+                await CheckReferences(nameof(_configuredProject.Services.ComReferences), _configuredProject.Services.ComReferences);
+                await CheckReferences(nameof(_configuredProject.Services.PackageReferences), _configuredProject.Services.PackageReferences);
+                await CheckReferences(nameof(_configuredProject.Services.ProjectReferences), _configuredProject.Services.ProjectReferences);
+                await CheckReferences(nameof(_configuredProject.Services.SdkReferences), _configuredProject.Services.SdkReferences);
+                await CheckReferences(nameof(_configuredProject.Services.WinRTReferences), _configuredProject.Services.WinRTReferences);
+
+                // UpToDateCheckInput is the special item group for customized projects to add explicit inputs
+                var upToDateCheckInputItems = project.GetItems("UpToDateCheckInput").Select(file => file.GetMetadataValue("FullPath"));
+
+                Log("Checking item type 'UpToDateCheckInput'.");
+                foreach (var upToDateCheckInputItem in upToDateCheckInputItems)
+                {
+                    Log($"Input item path '{upToDateCheckInputItem}'.");
+                    inputs.Add(upToDateCheckInputItem);
+                }
+
+                var latestInput = GetLatestTimestamp(inputs);
+                Log($"Lastest write timestamp on input is {latestInput}.");
+
+                var outputs = new List<string>();
+
+                outputs.AddRange(project.GetItems("UpToDateCheckOutput")
+                    .Select(file => file.GetMetadataValue("FullPath")));
+
+                IOutputGroupsService outputGroupsService = _configuredProject.Services.OutputGroups;
+                foreach (var outputGroup in KnownOutputGroups)
+                {
+                    Log($"Checking known output group {outputGroup}.");
+
+                    foreach (var output in (await outputGroupsService.GetOutputGroupAsync(outputGroup, cancellationToken)).Outputs)
```

Comment:
The same, if the outputGroupAsync doesn't cache any value, it can start an extra design time build. It can easily slow the product, instead of making it faster.